### PR TITLE
Fix GH-1146: Add server side validation for terms-conditions checkbox

### LINKED ISSUE #1146: "Agree terms" validation must be done on back end side
Rtmedia > Other settings > Ask user to agree terms 

image

Step to generate this scenario ->
Now delete the element `<div id="rtm-file_upload-ui" class="rtm-tab-content">` using any debugging tool
and then upload the file. 
image

It would allow user to upload file without checking the agree option.

### PR DESCRIPTION
**Pull request by SID177**
Add server side validation, before file upload and before activity is saved.
Fixes #1146

### CODE REVIEW

File: app/main/controllers/upload/RTMediaUploadEndpoint.php [hunk 1]
```diff
@@ -46,6 +46,38 @@ public function template_redirect( $create_activity = true ) {
...
+			// When activity upload terms are enabled on media upload page, we check whether someone has removed the html element or not.
+			if ( ! empty( $rtmedia->options['general_enable_upload_terms'] ) ) {
+
+				$uploader_request = wp_unslash( filter_input( INPUT_POST, 'uploader_terms_condition_request', FILTER_SANITIZE_STRING ) );
+				// We check for all other contexts, group and profile is checked above.
+				if ( ! empty( $uploader_request ) ) {
+					$terms_condition = wp_unslash( filter_input( INPUT_POST, 'uploader_terms_condition', FILTER_SANITIZE_STRING ) );
+					if ( empty( $terms_condition ) ) {
+						// This will be uploaded by JavaScript only so we send json response.
+						echo wp_json_encode( esc_html__( 'Terms and Conditions checkbox not found!', 'buddypress-media' ) );
```
Comment (abhijitrakas) [inline app/main/controllers/upload/RTMediaUploadEndpoint.php]: Use `wp_send_json_error`.

File: app/main/controllers/upload/RTMediaUploadEndpoint.php [hunk 2]
```diff
@@ -46,6 +46,38 @@ public function template_redirect( $create_activity = true ) {
...
+						die;
```
Comment (abhijitrakas) [inline app/main/controllers/upload/RTMediaUploadEndpoint.php]:
```suggestion
						
```
Then we don't need this.

File: app/main/controllers/upload/RTMediaUploadEndpoint.php [hunk 3]
```diff
@@ -46,6 +46,38 @@ public function template_redirect( $create_activity = true ) {
...
+			// When activity upload terms are enabled on activity page, we check whether someone has removed the html element or not.
+			if ( ! empty( $rtmedia->options['activity_enable_upload_terms'] ) ) {
+
+				// When media comment is uploaded, comment_media_activity_id will be there, so we check if it's not there.
+				$activity_request = wp_unslash( filter_input( INPUT_POST, 'activity_terms_condition_request', FILTER_SANITIZE_STRING ) );
+				if ( ! empty( $activity_request ) ) {
+					$terms_condition = wp_unslash( filter_input( INPUT_POST, 'activity_terms_condition', FILTER_SANITIZE_STRING ) );
+					if ( empty( $terms_condition ) ) {
+						// This will be uploaded by JavaScript only so we send json response.
+						echo wp_json_encode( esc_html__( 'Terms and Conditions checkbox not found!', 'buddypress-media' ) );
```
Comment (abhijitrakas) [inline app/main/controllers/upload/RTMediaUploadEndpoint.php]: Use `wp_send_json_error`.

File: app/main/controllers/activity/RTMediaBuddyPressActivity.php [hunk 4]
```diff
@@ -437,11 +437,34 @@ public function bp_get_activity_parent_content( $content ) {
...
+				} elseif ( isset( $class_obj->component ) ) {
+					// If error object is not found, we set component to null so it'll show error.
+					$class_obj->component = null;
+				}
+
+				if ( isset( $class_obj->error_type ) ) {
+					$class_obj->error_type = 'wp_error';
+				}
+
+				return '';
```
[the remark this answers is not in the document]
Comment (SID177) [reply app/main/controllers/activity/RTMediaBuddyPressActivity.php]: This is a filter, shouldn't we return something?

[the remark this answers is not in the document]
Comment (abhijitrakas) [reply app/main/controllers/activity/RTMediaBuddyPressActivity.php]: Ok, keep it as it is.

File: app/assets/js/rtm-upload-terms.js [hunk 5]
```diff
@@ -6,43 +7,93 @@
...
+        var terms_key = false;
+        var isTermsEnabled = false;
 
-        if ( 'false' === activity_terms_enabled ) {
+        if ( 'activity' === args.src ) {
+            request_key = 'activity_terms_condition_request';
+            terms_key = 'activity_terms_condition';
 
-            return false;
+            if ( 'true' === rtmedia_upload_terms_data.activity_terms_enabled ) {
```
Comment (abhijitrakas) [inline app/assets/js/rtm-upload-terms.js]: `rtmedia_upload_terms_data.activity_terms_enabled` this return `string/boolean` ?

File: app/assets/js/rtm-upload-terms.js [hunk 6]
```diff
@@ -6,43 +7,93 @@
...
+    } );
+
+    /**
+     * Check for the terms checkbox to be checked before the media is uploaded.
+     * Show warning if unchecked else proceed.
+     */
+    rtMediaHook.register( 'rtmedia_js_upload_file', function ( args ) {
+
+        if ( false === args || 'undefined' === typeof args.src ) {
+
```
Comment (abhijitrakas) [inline app/assets/js/rtm-upload-terms.js]:
```suggestion
```

File: app/assets/js/rtm-upload-terms.js [hunk 7]
```diff
@@ -6,43 +7,93 @@
...
     /**
-     * Check for the terms checkbox to be checked before the media is uploaded.
-     * Show warning if unchecked else proceed.
+     * Check terms condition checkbox before uploading files.
      */
-    rtMediaHook.register( 'rtmedia_js_upload_file', function ( args ) {
+    rtMediaHook.register( 'rtmedia_js_before_upload', function ( args ) {
 
-        if ( false === args ) {
+        if ( false === args || 'undefined' === typeof args.uploader.settings.multipart_params || 'undefined' === typeof args.src ) {
```
Comment (abhijitrakas) [inline app/assets/js/rtm-upload-terms.js]:
```suggestion
```

File: app/assets/js/rtm-upload-terms.js (hunk 5 above)
Comment (SID177) [reply app/assets/js/rtm-upload-terms.js]: It's localized so it'll be a string

File: app/assets/js/rtm-upload-terms.js [hunk 8]
```diff
@@ -6,43 +7,93 @@
...
  */
 if ( 'object' === typeof rtMediaHook ) {
+
     /**
-     * Check for the terms checkbox to be checked before the media is uploaded.
-     * Show warning if unchecked else proceed.
+     * Check terms condition checkbox before uploading files.
      */
-    rtMediaHook.register( 'rtmedia_js_upload_file', function ( args ) {
+    rtMediaHook.register( 'rtmedia_js_before_upload', function ( args ) {
```
Comment (abhijitrakas) [inline app/assets/js/rtm-upload-terms.js]: Add param doc.

File: app/assets/js/rtm-upload-terms.js [hunk 9]
```diff
@@ -6,43 +7,93 @@
...
+        args.uploader.settings.multipart_params = multipart_params;
+
+        return true;
+    } );
+
+    /**
+     * Check for the terms checkbox to be checked before the media is uploaded.
+     * Show warning if unchecked else proceed.
+     */
+    rtMediaHook.register( 'rtmedia_js_upload_file', function ( args ) {
```
Comment (abhijitrakas) [inline app/assets/js/rtm-upload-terms.js]: Add param doc.

**Pull request merged**